cpu/stm32/periph/spi_u5.c

```diff
@@ -0,0 +1,386 @@
+  periph_clk_dis(spi_config[bus].apbbus, spi_config[bus].rccmask);
+#ifdef STM32_PM_STOP
+  /* unblock STOP mode */
+  pm_unblock(STM32_PM_STOP);
+#endif
+  mutex_unlock(&locks[bus]);
+}
+
+static inline void _wait_for_end(spi_t bus) {
+  /* make sure the transfer is completed before continuing, see reference
+   * manual(s) -> section 'Disabling the SPI' */
+  /* When TSIZE>0, EOT and TXC signals are equal so polling of EOT is reliable
+  at whatever SPI communication mode to check end of the bus activity */
+  while (!(dev(bus)->SR & SPI_SR_EOT)) {
+  }
+  while (dev(bus)->SR & SPI_SR_EOT) {
+    dev(bus)->IFCR = SPI_IFCR_EOTC;
+  }
+  dev(bus)->IFCR = SPI_IFCR_TXTFC;
+}
+
+static void _transfer_no_dma(spi_t bus, const void *out, void *in, size_t len) {
+  // data to be sent tx
+  const uint8_t *outbuf = out;
+  // dest for data to receive rx
+  uint8_t *inbuf = in;
+
+  // lower 16 bits = TSIZE
+  // When these bits are changed by software, the SPI must be disabled. SPE = 0
+  dev(bus)->CR2 = (int16_t)len;
+
+  /* we need to recast the data register to uint_8 to force 8-bit access */
+  volatile uint8_t *TXDR = (volatile uint8_t *)&(dev(bus)->TXDR);
+  volatile uint8_t *RXDR = (volatile uint8_t *)&(dev(bus)->RXDR);
+
+  /*To restart the internal
+    state machine properly, SPI is strongly suggested to be disabled and
+    re-enabled before next transaction starts despite its setting is not
+    changed.*/
+  disable_spi(bus);
+  enable_spi(bus);
+
+  if (dev(bus)->CFG2 & SPI_CFG2_MASTER) {
+    assume(dev(bus)->CR1 & SPI_CR1_SPE);
+    // CSTART: master transfer start This bit can be set by software if SPI is
+    // *enabled* only
+    dev(bus)->CR1 |= SPI_CR1_CSTART;
+    assume(dev(bus)->CR1 & SPI_CR1_CSTART);
+  }
+
+  /* transfer data, use shortpath if only sending data */
+  if (!inbuf) {
+    for (size_t i = 0; i < len; i++) {
+      while (!(dev(bus)->SR & SPI_SR_TXP)) {
+      }
+      *TXDR = outbuf[i];
+      while ((dev(bus)->SR & SPI_SR_RXP)) { /* busy wait */
+                                            // discard incoming bytes
+        (void)*RXDR;
+      }
+    }
+  } else if (!outbuf) {
+    for (size_t i = 0; i < len; i++) {
+      while (!(dev(bus)->SR & SPI_SR_TXP)) { /* busy wait */
+      }
+      *TXDR = 0;
+      // The RXP is cleared as soon as less data than complete single packet is
+      // available in the RxFIFO
+      while (!(dev(bus)->SR & SPI_SR_RXP)) { /* busy wait */
+      }
+      inbuf[i] = *RXDR;
+    }
+  } else {
+    volatile uint16_t tx_remainder = len, rx_remainder = len;
+
+    while ((tx_remainder > 0 || rx_remainder > 0)) {
+      if (tx_remainder > 0 && dev(bus)->SR & SPI_SR_TXP) {
+        *TXDR = outbuf[len - tx_remainder];
+        tx_remainder--;
+      }
+      if (rx_remainder > 0) {
+        if (!(dev(bus)->SR & SPI_SR_RXP))
+          continue;
+        inbuf[len - rx_remainder] = *RXDR;
+        rx_remainder--;
+      }
+      if (dev(bus)->SR & SPI_SR_EOT) {
+        break;
```

Comment:
In my testing on an STM32H7, this break causes the last received bytes to be left in the RX FIFO. The transfer completes (ie the hardware has shifted in/out the requested number of bytes), and we break out of the loop, leaving `rx_remainder != 0`. Is there a need for this check or another way it could be done?

Reply:
This explains my comment above:

> On my H7. It seems the driver is reading 1 byte shy of what is requested I have to transfer 3 bytes to get 2 bytes to be modified in the rx buffer. Still digging into why that is.

Reply:
Does `dev(bus)->SR` indicate that there might be some bytes left? 

Does this help?
```suggestion
      if (dev(bus)->SR & SPI_SR_EOT && rx_remainder == 0) {
        break;
```

Reply:
I've just ran into that issue myself while using the the fully featured SPI1 interface, the errata mentions an possible truncation issue however I do not understand how the mentioned (below) issue would apply to my PR since the SPI is only disabled once all bytes are received.

> After an EOT event signaling the end of a non-zero transfer size transaction (TSIZE > 0) upon sampling the last
data bit, the software may disable the SPI peripheral. As expected, disabling SPI deactivates the SPI outputs
(SCK, MOSI and SS when the SPI operates as a master, MISO when as a slave), by making them float or
statically output their by-default levels, according to the AFCNTR bit of the SPI_CFG2 register.
With fast software execution (high PCLK frequency) and slow SPI (low SCK frequency), the SPI disable occurring
too fast may result in truncating the SPI output signals. For example, the device operating as a master then
generates an asymmetric last SCK pulse (with CPHA = 0), which may prevent the correct last data bit reception
by the other node involved in the communication.
Workaround
Apply one of the following measures or their combination:
• Decrease the ratio between PCLK and SCK frequencies.
• Add a delay between the EOT event and SPI disable action

Reply:
It seems like the SPI peripheral looses some rx bits only if `TSIZE` is set on an fully featured interface. Not setting `TSIZE` seems to fix the problem. Tested with an sd card i9n SPI mode.